File: 320 Generalized Abbreviation.java
import java.util.*;
/*
320 Generalized Abbreviation
https://leetcode.com/problems/generalized-abbreviation/
*/
public class Solution {
    
    private ArrayList<String> ans = new ArrayList<String>();
    private String word = "";
    
    public void rec(StringBuffer sb, int pos){
        int len = sb.length();
        if(pos >= word.length()){
            ans.add(sb.toString());
        } else if (pos == word.length() - 1){
            // Optimization
            sb.append(word.charAt(pos));
            ans.add(sb.toString());
            sb.delete(len,sb.length());
            sb.append(1);
            ans.add(sb.toString());
            sb.delete(len,sb.length());
        } else {
            for(int i = 1;i < word.length() - pos;i++){
                sb.append(i);
                sb.append(word.charAt(pos+i));
                rec(sb,pos+i+1);
                sb.delete(len,sb.length());
            }
            
            // w, 1 => w3
            sb.append(word.length() - pos);
            ans.add(sb.toString());
            sb.delete(len,sb.length());
            
            // w, 1 => wo, 2
            sb.append(word.charAt(pos));
            rec(sb,pos+1);
            sb.delete(len,sb.length());
        }
    }
    
    public List<String> generateAbbreviations(String word) {
        this.ans.clear();
        this.word = word;
        StringBuffer sb = new StringBuffer();
        rec(sb,0);
        return this.ans;
    }
}

File: 320 Generalized Abbreviation.js
/**
 * @param {string} word
 * @return {string[]}
 */
var generateAbbreviations = function(word) {
    var i,j,c;
    var size = word.length;
    var width = 1 << size;
    
    var res = [];
    var ls = [];
    for(i=0;i<width;i++){
        ls = [];
        c = 0;
        for(j=0;j<size;j++){
            if(1 << j & i){
                if(c) ls.push(c.toString());
                c = 0;
                ls.push(word[j]);
            } else {
                c++;
            }
        }
        if(c) ls.push(c);
        res.push(ls.join(''));
    }
    return res;
};
